src/cheri_pte.sail

```diff
@@ -42,98 +42,62 @@ union PTE_Check = {
-                                       then {
-                                         if   e.StoreCap() == 0b1
-                                         then (true,  PTE_CAP_OK)
-                                         else (false, PTE_STORE_CAP_ERR)
-                                       }
-                                       else (false, PTE_CAP_OK)
-                                     },
-
-    (ReadWrite(Cap),  Supervisor) => { let e = Mk_Ext_PTE_Bits(ext);
-                                       if   (p.U() == 0b0 | do_sum) & p.R() == 0b1 & p.W() == 0b1
-                                       then {
-                                         if   e.StoreCap() == 0b1 & e.LoadCap() == 0b1
-                                         then (true, PTE_CAP_OK)
-                                         else if e.StoreCap() == 0b0
-                                         /* return a failure since we should cause an exception */
-                                         then (false, PTE_STORE_CAP_ERR)
-                                         /* return a success since the translation should proceed */
-                                         else (true, PTE_LOAD_CAP_ERR)
-                                       }
-                                       else (false, PTE_CAP_OK)
-                                     },
-
-    (_,               Machine)    => internal_error("m-mode mem perm check")
+    (Read(_),      User)       => p.U() == 0b1 & (p.R() == 0b1 | (p.X() == 0b1 & mxr)),
+    (Write(_),     User)       => p.U() == 0b1 & p.W() == 0b1,
+    (ReadWrite(_), User)       => p.U() == 0b1 & p.W() == 0b1 & (p.R() == 0b1 | (p.X() == 0b1 & mxr)),
+    (Execute(),    User)       => p.U() == 0b1 & p.X() == 0b1,
+
+    (Read(_),      Supervisor) => (p.U() == 0b0 | do_sum) & (p.R() == 0b1 | (p.X() == 0b1 & mxr)),
+    (Write(_),     Supervisor) => (p.U() == 0b0 | do_sum) & p.W() == 0b1,
+    (ReadWrite(_), Supervisor) => (p.U() == 0b0 | do_sum) & p.W() == 0b1 & (p.R() == 0b1 | (p.X() == 0b1 & mxr)),
+    (Execute(),    Supervisor) => p.U() == 0b0 & p.X() == 0b1,
+
+    (_,            Machine)    => internal_error("m-mode mem perm check")
+  };
+
+  let e = Mk_Ext_PTE_Bits(ext);
+  let (succ, pte_chk) : (bool, pte_check) =
+  match (base_succ, ac) {
+    /* Base translation exceptions take priority over CHERI exceptions */
+    (false, _)               => (false, PTE_CAP_OK),
+
+    (true,  Read(Cap))       => if   e.LoadCap() == 0b1
+                                then (true, PTE_CAP_OK)
+                                /* Allow the address translation to proceed, but mark for tag stripping */
+                                else (true, PTE_LOAD_CAP_ERR),
+
+    (true,  Write(Cap))      => if   e.StoreCap() == 0b1
+                                then (true, PTE_CAP_OK)
+                                /* Do not allow the address translation to proceed */
+                                else (false, PTE_STORE_CAP_ERR),
```

Comment:
#18 considered this OK and left the caveating to `cheri_insts` for data-dependence's sake.